apps/OboeTester/app/src/main/cpp/ReverseJniEngine.cpp

```diff
@@ -0,0 +1,127 @@
+
+    // Get the method ID for the onAudioReady callback in the Java class.
+    // The signature "([FII)V" means a method that takes a float array and two ints, and returns void.
+    jclass javaClass = env->GetObjectClass(mJavaObject);
+    mOnAudioReadyId = env->GetMethodID(javaClass, "onAudioReady", "([FII)V");
+    env->DeleteLocalRef(javaClass);
+}
+
+ReverseJniEngine::~ReverseJniEngine() {
+    // It's important to release the global reference when the engine is destroyed.
+    JNIEnv *env;
+    mJavaVM->GetEnv(reinterpret_cast<void **>(&env), JNI_VERSION_1_6);
+    env->DeleteGlobalRef(mJavaObject);
+}
+
+void ReverseJniEngine::start(int bufferSizeInBursts) {
+    LOGI("Starting audio stream.");
+    setupAudioStream();
+    setBufferSizeInBursts(bufferSizeInBursts);
+    if (mAudioStream) {
+        mAudioStream->requestStart();
+    }
+}
+
+void ReverseJniEngine::stop() {
+    __android_log_print(ANDROID_LOG_INFO, TAG, "Stopping audio stream.");
+    if (mAudioStream) {
+        mAudioStream->requestStop();
+        closeAudioStream();
+    }
+}
+
+void ReverseJniEngine::setBufferSizeInBursts(int bufferSizeInBursts) {
+    if (mAudioStream) {
+        mAudioStream->setBufferSizeInFrames(bufferSizeInBursts * mAudioStream->getFramesPerBurst());
+        LOGI("Buffer size set to %d frames.", mAudioStream->getBufferSizeInFrames());
+    }
+}
+
+void ReverseJniEngine::setupAudioStream() {
+    oboe::AudioStreamBuilder builder;
+    builder.setDirection(oboe::Direction::Output)
+            ->setPerformanceMode(oboe::PerformanceMode::LowLatency)
+            ->setSharingMode(oboe::SharingMode::Exclusive)
+            ->setFormat(oboe::AudioFormat::Float)
+            ->setChannelCount(oboe::ChannelCount::Mono)
+            ->setDataCallback(this);
+
+    oboe::Result result = builder.openStream(mAudioStream);
+    if (result != oboe::Result::OK) {
+        LOGE("Failed to create stream. Error: %s", oboe::convertToText(result));
+    }
+}
+
+void ReverseJniEngine::closeAudioStream() {
+    if (mAudioStream) {
+        mAudioStream->close();
+    }
+}
+
+oboe::DataCallbackResult ReverseJniEngine::onAudioReady(oboe::AudioStream *oboeStream,
+                                                        void *audioData,
+                                                        int32_t numFrames) {
+    JNIEnv *env;
+    // Attach the current thread to the JVM to make JNI calls.
+    int getEnvStat = mJavaVM->GetEnv(reinterpret_cast<void **>(&env), JNI_VERSION_1_6);
+    if (getEnvStat == JNI_EDETACHED) {
+        if (mJavaVM->AttachCurrentThread(&env, nullptr) != 0) {
+            LOGE("Failed to attach current thread to JVM");
+            return oboe::DataCallbackResult::Stop;
+        }
+    }
+
+    int xRunCount = mAudioStream->getXRunCount().value();
+
+    // Create a new Java float array to pass the audio data.
+    jfloatArray audioDataArray = env->NewFloatArray(numFrames * mAudioStream->getChannelCount());
```

Comment:
Allocating the array can be expensive if it is done per callback. Could it be faster by set the callback size and allocate the data in JAVA? In this case, every callback is asking for a fixed size data and can be always fulfill by JAVA side and read from callback thread. In that case, there is no need to allocate here.